p2p: Drop unsolicited CMPCTBLOCK from non-HB peer and when blocksonly

davidgumberg opened the pull request:
Processing unsolicited `CMPCTBLOCK`'s from a peer that has not been marked high bandwidth is not well-specified behavior in BIP-0152, in fact the BIP seems to imply that it is not permitted:

> "[...] method is not useful for compact blocks because `cmpctblock` blocks can be sent unsolicitedly in high-bandwidth mode"

See https://github.com/bitcoin/bips/blob/master/bip-0152.mediawiki#separate-version-for-segregated-witness

This PR disables processing of CMPCTBLOCK messages in three cases:
$1$. When the block is unsolicited and from a non-HB peer.
$2$. When this node is running in `-blocksonly` mode.
$3$. When the peer has not advertised `CMPCTBLOCK` support with a `SENDCMPCT` message.

Not processing unsolicited blocks slightly raises the cost of discovering a peer's mempool via `CMPCTBLOCK` as described in #28272. As pointed out there, getting an HB slot is relatively easy, so this does not prevent an attacker from doing this, it just slightly raises the bar.

Probably more important is not processing `CMPCTBLOCK` messages as a `-blocksonly` node. A blocksonly node has a lot less surface area for leaking its mempool since it does no transaction relay, and leaking a blocksonly node's mempool is pretty dangerous since it is very likely to be the origin for all of the transactions in its mempool.

davidgumberg committed:
p2p: Drop unsolicited CMPCTBLOCK from non-HB peer

Processing unsolicited CMPCTBLOCK's from a peer that has not been marked
high bandwidth is not well-specified behavior in BIP-0152, in fact the
BIP seems to imply that it is not permitted:

"[...] method is not useful for compact blocks because `cmpctblock`
blocks can be sent unsolicitedly in high-bandwidth mode"

See https://github.com/bitcoin/bips/blob/master/bip-0152.mediawiki#separate-version-for-segregated-witness

This partially mitigates a mempool leak described in
[#28272](https://github.com/bitcoin/bitcoin/issues/28272), but that
particular issue will persist for peers that have been selected as high
bandwidth.

This also mitigates potential DoS / bandwidth-wasting / abusive
behavior that is discussed in the comments of #28272.

davidgumberg committed:
test: p2p: Nodes ignore unsolicited CMPCTBLOCK's

mzumsande reviewed (commented):
Concept ACK

In the commit message:
"making it only possible for peers selected as high bandwidth to discover a node's
mempool" is a bit too strong, since mempools are discoverable, just not  transactions that were added to them after the last inv to the peer.

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/32606/commits/1fc95e080040a9d038a8291eeecf645e0413a5c8

Crypt-iQ commented on src/net_processing.cpp:
```diff
@@ -4435,6 +4435,11 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
             range_flight.first++;
         }
 
+        if (!requested_block_from_this_peer && !pfrom.m_bip152_highbandwidth_to) {
```
I think this check is still permissive enough such that a `NetMsgType::CMPCTBLOCK` message can be sent to a node in -blocksonly mode and still get processed. This would require the -blocksonly node to request the block from the peer.

Maybe something like this instead?
```diff
+.       if ((!requested_block_from_this_peer && !pfrom.m_bip152_highbandwidth_to) || m_opts.ignore_incoming_txs) {
```

Crypt-iQ replied:
> One could special-case it to account for allowing empty block reconstruction on blocks-only nodes but it feels like dead weight in the code.

Yeah, I feel like the complexity isn't worth it.

ajtowns replied:
> I think this check is still permissive enough such that a `NetMsgType::CMPCTBLOCK` message can be sent to a node in -blocksonly mode and still get processed. T

Should we only send `SENDCMPCT hb=false` after `VERACK` if we're not blocksonly? Making all the compact block stuff conditional on not being `-blocksonly=1` would make sense to me, but probably isn't needed for this PR.

polespinasa replied:
What if in the case of being in `-blocksonly` mode and receive a `CMCTBLOCK` message we instantly respond with a `GETBLOCKTXN` asking for all transactions in case the block is not empty?

Crypt-iQ replied:
Agree that it doesn't make sense to process, we should drop it asap. Haven't tested either of the below things:

`ProcessHeadersMessage` will trigger a `GETDATA` and if the block turns out to be invalid, our peer who sent us the `CMPCTBLOCK` won't respond and will get disconnected for stalling. If our peer is modified to relay these in the first place to all peers, they could isolate themselves from -blocksonly nodes (assuming they aren't a mining node and don't validate what they send out). This is https://github.com/bitcoin/bitcoin/issues/13370.

A tangent, but I think a miner could trigger this statement in honest peer's connections if they mined and withheld enough blocks (~4-5) _and_ if the last `CMPCTBLOCK` is for an invalid block. I'm leaving out a few steps and can write out what I think on the linked issue if it's worth addressing?

Crypt-iQ reviewed (commented):
My main concern here is if this disables a FIBRE future; FIBRE apparently used unsolicited compact blocks (see: https://github.com/bitcoin/bitcoin/issues/28272#issuecomment-2905778748). If we are not concerned about that, then Concept ACK.

ajtowns commented:
> My main concern here is if this disables a FIBRE future; FIBRE apparently used unsolicited compact blocks

The fact that you're enabling FIBRE on a connection can just be taken as implicitly soliciting compact blocks on that connection, so I don't think this is a problem.

hodlinator reviewed (commented):
Concept ACK 1fc95e080040a9d038a8291eeecf645e0413a5c8

Crypt-iQ commented:
> The fact that you're enabling FIBRE on a connection can just be taken as implicitly soliciting compact blocks on that connection, so I don't think this is a problem.

Right, my earlier comment was wrong.

I think that dropping unsolicited cmpctblock in the non -blocksonly case is kind of ineffective as pointed out by others. But I also don't have a strong opinion about it given that it doesn't affect FIBRE.

ajtowns commented on src/net_processing.cpp:
```diff
@@ -4435,6 +4435,11 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         }
 
+        if (!requested_block_from_this_peer && !pfrom.m_bip152_highbandwidth_to) {
+            LogDebug(BCLog::NET, "Peer %d, not marked as high-bandwidth, sent us an unsolicited compact block!\n", pfrom.GetId());
```
This doesn't seem to have been taken (or has been lost in a rebase?)

hodlinator commented on test/functional/p2p_compactblocks.py:
```diff
@@ -950,6 +968,63 @@ def announce_cmpct_block(node, peer, txn_count):
+        def assert_highbandwidth_states(node, idx, hb_to, hb_from=False):
+            peerinfo = node.getpeerinfo()[idx]
+            assert_equal(peerinfo['bip152_hb_to'], hb_to)
+            assert_equal(peerinfo['bip152_hb_from'], hb_from)
```
Duplicated from other test in same file, could be made class-local instead?
(Last arg is always left at default in this new outer test function).

hodlinator commented on test/functional/p2p_compactblocks.py:
```diff
@@ -950,6 +968,63 @@ def announce_cmpct_block(node, peer, txn_count):
+
+        block, msg = build_compact_block()
+        unsolicited_peer.send_without_ping(msg_headers([block]))
+        unsolicited_peer.wait_for_getdata([block.hash_int], timeout=30)
```
The reason we aren't marked as HB peer here is that the block hash was already known from previous messages, right? Maybe worth pointing out with a comment?

Could also use `assert_highbandwidth_states(node, idx=-1, hb_to=False)` another time at the end of this block to verify.

Crypt-iQ replied:
`unsolicited_peer` also hasn't sent `sendcmpct` so its `m_provides_cmpctblocks` is false, and the node still processes the compact block! I think it is useful to have coverage for this case.

davidgumberg replied:
Messing around with this I think there's actually a bug, even if `unsolicited_peer` sends a `sendcmpct`, they don't get a high bandwidth slot when they should, since they were the first to announce a block.

I believe this is because mapBlocksInFlight doesn't get cleared when a node reconstructs from a `GETBLOCKTXN` resulting in `MaybeSetPeerAsAnnouncing.*` never getting called:

https://github.com/bitcoin/bitcoin/blob/9e4567b17a285a266fa540b4141e0db5781cc228/src/net_processing.cpp#L2191-L2197

I will change this test so that a `sendcmpct` is sent, and introduce a fix for @Crypt-iQ's suggestion that we shouldn't process blocks from peers where `m_provides_cmpctblocks` is false, and try to fix this behavior in a follow-up.

davidgumberg replied:
Yeah in order to perform the attack on nodes that have >=1 honest HB peer, the malicious peer has to be able to produce valid PoW block headers for blocks no one has. Maybe this is not irrelevant since someone doing selfish mining (and not trying to hide it) could probably slightly extend their advantage by broadcasting headers of unpublished blocks and then stalling.

hodlinator reviewed (commented):
Reviewed 5caaefd8d4174cfc27d5538e7f2d3a1c2bdb77d8

Crypt-iQ commented on test/functional/p2p_compactblocks.py:
```diff
@@ -265,9 +265,13 @@ def check_announcement_of_new_block(node, peer, predicate):
 
     # This test actually causes bitcoind to (reasonably!) disconnect us, so do this last.
     def test_invalid_cmpctblock_message(self):
+        self.make_peer_hb_to_candidate(self.nodes[0], self.segwit_node)
```
Since `self.segwit_node` was disconnected in the prior test it won't be marked as HB. I agree that we should assert that it is marked as HB though.

davidgumberg replied:
@hodlinator's suggested change passes, because segwit node's index is *not* `0` any more, it just got connected so it's `-1`, `0` is some other node that's already HB. And the test passes without making segwit_node HB since I've modified the test to actually be solicited, since the peer sends a header and waits for getdata. Should I add separate test cases for DC'ing both unsolicited cmpctblock's from HB's and solicited compact blocks from non-HB's?

It also seems weird that we'll accept a CMPCTBLOCK from someone that hasn't given us a SENDCMPCT announcing support for it, but that's probably fine, I'm thinking out loud but maybe there's some way in which that's an issue if later versions of the CMPCTBLOCK message are introduced in the future.

In my opinion, the statefulness of `p2p_compactblocks.py` in general is pretty awful and would be worth fixing in a separate PR.

hodlinator replied:
Thanks for using separate peers in 42e281dba0f0807cc08d8c50a9d19906e27e2129. Was totally not expecting the reassignment of `segwit_node`. :sleeping:

I like how you use `hb_peer_idx` in f24237ef3d8c94c01f4b9f86a7d28d0b5fcdc75b as both a parameter for `assert_highbandwidth_states()` and for indexing into `self.nodes[0].p2ps`.

Crypt-iQ reviewed (commented):
I agree with @dergoegge that it's easy to become a HB peer if you're already the first to send unsolicited compact blocks to try and fingerprint a node's mempool. I think the benefit with this change is being more strict with what we'll accept and following the protocol closer (and I think there's even more improvements we can make). Left some test-related nits.

davidgumberg committed:
refactor: test: Static assert_highbandwidth_states

This is move-only, and allows assert_highbandwidth_states to be used by
other tests.

davidgumberg committed:
test: (Un)solicited invalid cb -> get disconnected.

Modifies the invalid_cmpctblock_message test to check that both HB peers
sending unsolicited and non-HB peers sending solicited invalid
cmpctblock's get the boot from us. Also refactors the test to make it
less stateful.

instagibbs commented:
@Crypt-iQ `BlockEncoding.*` benchmarks should cover this up to 50k. Quickly tested 100k, and it seems ~linear slowdown. (~4ms on my mid laptop)

ajtowns commented:
> the second peer can endlessly send CMPCTBLOCK's

If we're concerned about that, seems better to fix it directly?

```diff
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -4412,6 +4412,11 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         while (range_flight.first != range_flight.second) {
             if (range_flight.first->second.first == pfrom.GetId()) {
                 requested_block_from_this_peer = true;
+                const auto& qb_list_it = range_flight.first->second.second;
+                if (qb_list_it->partialBlock != nullptr) {
+                    LogDebug(BCLog::CMPCTBLOCK, "Ignoring duplicate CMPCTBLOCK message for block %s received from peer=%d%s", pindex->GetBlockHash().ToString(), pfrom.GetId(), pfrom.LogIP(fLogIPs));
+                    return;
+                }
                 break;
             }
             range_flight.first++;
```

> But, as I see it, there is very https://github.com/bitcoin/bitcoin/pull/32606#issuecomment-3457245344 a legitimate/useful reason today to process an unsolicited CMPCTBLOCK message

I could imagine someone restarting the FIBRE network, and having it interface to the p2p network with a custom node implementation that implements logic like "only relay txs with an internal node. drop connections to anyone with a ping time exceeding 50ms. for each block, did I get this block via FIBRE? if so, send a CMPCTBLOCK message to all peers because I'm probably the quickest source in this region. if not, behave normally."

davidgumberg commented:
> If we're concerned about that, seems better to fix it directly?
> 
> ```diff
> --- a/src/net_processing.cpp
> +++ b/src/net_processing.cpp
> @@ -4412,6 +4412,11 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
>   [...]
> ```

The suggested diff doesn't work because we remove the block request from mapBlocksInFlight if the txn's are missing, the `firstInFlight` slot is already taken, and the block is not from an HB Peer:

https://github.com/bitcoin/bitcoin/blob/24434c1284b84e2c33cf3240ca677d77cad298b2/src/net_processing.cpp#L4494-L4497

To do something like what you're suggesting I think we would need another map that tracks whether or not a peer has already had the block in flight at one point and we gave up on it.

> I could imagine someone restarting the FIBRE network, and having it interface to the p2p network with a custom node implementation that implements logic like "only relay txs with an internal node. drop connections to anyone with a ping time exceeding 50ms. for each block, did I get this block via FIBRE? if so, send a CMPCTBLOCK message to all peers because I'm probably the quickest source in this region. if not, behave normally."

If I'm not mistaken, at least the only effort I know of at restarting the FIBRE network does not want to reveal the identity of the FIBRE nodes to prevent targeted DoS, but maybe I am mistaken, pinging @w0xlt.

But, in general, I hear the broader stroke of your objection as: The "hole" in our implementation of compact block relay might actually be an opportunity we can leverage to improve block propagation, and we shouldn't close it so hastily. And that's fair enough, I kind of disagree about the FIBRE use case, but it is worth thinking about if, in an implementation like you suggested, where each peer only gets one unsolicited block per valid header, the price of `InitData()` * number of connections an attacker can reasonably get to a node is worth whatever the potential improvement we could get from nodes that definitely know they've got the block first like mining nodes, or are reasonably confident they're the fastest like FIBRE nodes.

ajtowns commented:
> The suggested diff doesn't work because we remove the block request from mapBlocksInFlight if the txn's are missing, the firstInFlight slot is already taken, and the block is not from an HB Peer:

Does that logic make sense? If we receive:

```mermaid
sequenceDiagram
   participant A
   Us ->> A: SENDCMPCT 0
   Us ->> B: SENDCMPCT 1 (high-bandwidth)
   A ->> Us: t=0 HEADERS
   Us ->> A: GETDATA(CMPCT)
   B ->> Us: t=25ms CMPCTBLOCK
   Us ->> B: GETBLOCKTXN
   A ->> Us: t=50ms CMPCTBLOCK
```

does it really make sense to not request the missing txns from the peer who first announced the header to us? If the RTT latency between us and B is 80ms (versus us and A at 50ms), then we'd get a reply from A at t=100ms and the reply from B at t=105ms, eg.

Not requesting GETBLOCKTXN in response to a CMPCTBLOCK from a non-hb peer when we didn't request the cb and we already have a request in flight still sounds sensible, though, so I don't think this resolves the objection.

Finding a way to refactor this code so that the logic for how we do block relay is mostly in one place rather than scattered amongst message various different message parsing logic might be valuable... I guess if you treated the logic as:

 1. Ooh, this peer is telling me about a possibly new block header!
 2. Ooh, it would be a new tip and I don't have the block yet! And I don't have it in-flight yet, I should reque...
 3. Ooh, it's already a CMPCTBLOCK and I don't need to request it! Great, let's analyse it, see if it reconstructs!
 4. Oh, it doesn't, better request the transactions

then if you received an unsolicited CMPCTBLOCK message that you already had things in flight for that block, you'd stop at (2) and just treat the message as an overly verbose headers update, and not even try to process it against the mempool (after all, if you have requests in flight, that was probably from an honest peer and you probably are actually missing transactions, so the reconstruction would probably fail anyway). OTOH, if you didn't have things in flight for it (or weren't being as parallel as you'd like to be), you would continue, and would request (and eventually process) the missing transactions.

polespinasa reviewed (approved):
code review ACK  32bfd6136134e7194ea0b56ec08498f66829fc40

I think it's worth resolving the odd case for `-blocksonly` in this pull request. But I would be happy too with it being done in a follow-up PR.

Crypt-iQ reviewed (commented):
> treat a CMPCTBLOCK announcement as being fine at any time (eg, so that miners who get a block via RPC can just send CMPCTBLOCK messages directly), and the "high-bandwidth" as just a way for nodes to avoid wasting their bandwidth sending those messages unnecessarily

I think yes, except for -blocksonly nodes, where on receipt we wouldn't punish or anything, but we would ideally drop the message?

> change our behaviour when sending GETBLOCKTXN to also request any transactions that were in our mempool but had a sequence number higher than when we first saw the new block's header (If you completely reconstructed the block, but used newer txns, you would still not send a GETBLOCKTXN)

I think it would be possible for an attacker to send `HEADERS`, receive `GETDATA`, wait for some time to allow new txns to accumulate in the node's mempool, then send a `CMPCTBLOCK` with suspected new txns that wouldn't have been announced yet. But, if the targeted node receives the valid block from any other peer, the download state gets wiped. So this is an edge-of-an-edge case. Since an attacker wants "tight timing information" as you pointed out on the linked issue and this is only able to be done ~10mins (and may not even work all the time if the attacker is not HB), it doesn't seem urgent.

> without something concrete I would prefer to close down our implementation of the protocol to be as conservative as possible about what CMPCTBLOCK messages we process

Any restrictive change here introduces a kind of implicit version for compact block relay even if it's later lifted. I guess the unknown for me is whether this is something that miners (or others) want to do, are planning to do, or if this is a usecase we should support?

+1 to the refactor suggestion. I don't have any suggestions off-hand. ~~Maybe a new file could be introduced specifically for compact block relay (similar to what was done for blockencodings.cpp)? It would make writing tests a bit easier.~~ Also, checking whether `partialBlock` is nullptr (and having to reset it in some cases), is awkward.

hodlinator reviewed (commented):
Reviewed 32bfd6136134e7194ea0b56ec08498f66829fc40

Crypt-iQ reviewed (commented):
Reviewed 32bfd6136134e7194ea0b56ec08498f66829fc40 and left some comments. I think if the use-case brought up by @ajtowns is not a concern, then I can ACK.

Separately, I've been thinking about how a refactor would look. I'm still at a loss because unifying the headers and the compact blocks processing logic seems difficult because sometimes the compact block can be processed as a headers message (i.e. calls `ProcessHeadersMessage`) and there are compact blocks-specific checks that a headers message wouldn't go through.

w0xlt reviewed (commented):
reACK https://github.com/bitcoin/bitcoin/pull/32606/commits/32bfd6136134e7194ea0b56ec08498f66829fc40

davidgumberg commented on src/net_processing.cpp:
```diff
@@ -4412,6 +4412,11 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
 
+        if (!requested_block_from_this_peer && !pfrom.m_bip152_highbandwidth_to) {
+            LogDebug(BCLog::NET, "Peer %d, not marked as high-bandwidth, sent us an unsolicited compact block!\n", pfrom.GetId());
+            return;
```
@Crypt-iQ In this case, I don't think a message would be logged since we'll return above after the check that we have data for the block:

https://github.com/bitcoin/bitcoin/blob/9e4567b17a285a266fa540b4141e0db5781cc228/src/net_processing.cpp#L4600-L4602

davidgumberg committed:
p2p: make blocksonly nodes ignore CMPCTBLOCK messages

blocksonly nodes don't benefit from compact blocks, since they don't
have a mempool to aid in reconstruction, so they should not process
CMPCTBLOCK messages.

This is not just belt-and-suspenders, as a blocksonly node will
trivially reveal exactly which transactions in a block are its own in
the GETBLOCKTXN response to a CMPCTBLOCK. Since it will be missing every
transaction in the block, except for its own.

See discussion: https://github.com/bitcoin/bitcoin/issues/28272

davidgumberg committed:
p2p: Ignore CMPCTBLOCK from peer that hasn't sent SENDCMPCT

This commit also changes the default sendcmpct version in the functional
test to `2`, since this is the version that nodes expect, prior to this
commit, nowhere in the functional test framework was the default version
value used:

git grep -P 'msg_sendcmpct\((?![^)]*version\s*=)' HEAD^

`version=2` is a sensible default, since this is the version nodes
currently expect in the SENDCMPCT handshake.

davidgumberg commented on test/functional/p2p_compactblocks.py:
```diff
@@ -932,32 +934,40 @@ def announce_cmpct_block(node, peer, txn_count):
             assert_equal([peer['bip152_hb_to'] for peer in node.getpeerinfo()], [False, True, True, True])
 
-            block, cmpct_block = announce_cmpct_block(node, stalling_peer, num_missing)
+            block, cmpct_block = announce_cmpct_block(node, stalling_peer, num_missing, expect_getblocktxn=False)
```
Subtle, thanks for catching this, I've reworked this so that in the first commit where we drop CMPCTBLOCK messages that are unsolicited I change the test to send a headers, and in the later commit where I tighten down  the getblocktxn stuff, I dropped this `expect` change.

davidgumberg commented on src/net_processing.cpp:
```diff
@@ -4412,6 +4412,11 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
             range_flight.first++;
         }
 
+        if (!requested_block_from_this_peer && !pfrom.m_bip152_highbandwidth_to) {
```
I don't think a comment would add much more than the log message here.

davidgumberg commented on test/functional/p2p_compactblocks.py:
```diff
@@ -264,17 +264,37 @@ def check_announcement_of_new_block(node, peer, predicate):
         cmpct_block.header = CBlockHeader(block)
         cmpct_block.prefilled_txn_length = 1
         # This index will be too high
         prefilled_txn = PrefilledTransaction(1, block.vtx[0])
```
Taken, thanks.

hodlinator committed:
qa: Avoid UTXO reuse between test functions

hodlinator reviewed (commented):
Reviewed cb60b6e9ed85e2296c5bd792da4445aeff0f0b58

Made a suggestions branch: https://github.com/hodlinator/bitcoin/tree/pr/32606_suggestions [compare](https://github.com/bitcoin/bitcoin/compare/cb60b6e9ed85e2296c5bd792da4445aeff0f0b58...451cf047bf046ea16c4a15e5dfa5f68d51515c6a)

davidgumberg commented on test/functional/p2p_compactblocks.py:
Taken, thanks.

hodlinator reviewed (commented):
Reviewed b11753f6a6a985ad6b5d2890f28323d4d5d5582d

Crypt-iQ commented on test/functional/p2p_compactblocks_blocksonly.py:
```diff
@@ -122,5 +156,18 @@ def test_for_cmpctblock(block):
         self.nodes[0].submitblock(block2.serialize().hex())
         p2p_conn_blocksonly.wait_until(lambda: test_for_cmpctblock(block2))
 
+        # This is redundant with other tests, and is here as a test-of-the-test
```
nice addition 👍

Crypt-iQ reviewed (commented):
I like the added tests. I think these tests could be expanded and also cleaned up in the future since I think reusing the peer connections between sub-tests is a little confusing, but that's unrelated to this PR. My one critique is that the checks for -blocksonly and if the node has sent `sendcmpct` should be at the top of the scope imo. They are safe as-is, though I really do think any defensive checks should fail quickly. It prevents things like code accidentally being introduced between the top of the scope and the check, even if it's unlikely that it happens. Other than that, I think this is good to go. 

I'm not concerned about the earlier discussion about the use-case of unsolicited cmpctblocks being sent from non-HB peers. It doesn't seem like anybody has plans to implement that, though I'm happy to be proven wrong.

davidgumberg commented on src/net_processing.cpp:
```diff
@@ -4597,8 +4597,13 @@ void PeerManagerImpl::ProcessMessage(Peer& peer, CNode& pfrom, const std::string
+        // Nothing to do here
+        if (pindex->nStatus & BLOCK_HAVE_DATA) {
             return;
+        } else if (m_opts.ignore_incoming_txs) {
```
I'm not sure that I agree, I think when it comes to something that is expensive to produce like a high-PoW header,  and that  could be an escape valve for an e.g. eclipse attack (a high-PoW chaintip) I think we want to be pretty aggressive in pursuing it. 

But that is just a kind of rhetorical / intuitive argument, I'm not sure how to be more thorough in comparing the potential harm of processing a protocol-violating CMPCTBLOCK message to check for a high-PoW header vs the benefit of potentially learning about a high-PoW chain.

Crypt-iQ replied:
This doesn't make sense to me because it's protocol-invalid. The -blocksonly node should never accept cmpctblocks at all, it never requests them in the first place. What are the odds that we're eclipsed by not having the header (it's also possible to be eclipsed by having the header and not the block which this wouldn't help) and we have a protocol-violating peer who sends us a compact block to bail us out? I think the code should be very strict in what we accept, not doing this leads to edge cases like addrfetch connections being marked HB or processing compact blocks when sendcmpct was never sent. Anecdotally, I've found bugs in other projects where the root cause is being loose in what we accept.

Crypt-iQ replied:
There was more (inconclusive) discussion on [IRC](https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2026-03-26#1206397;). Sure, tightening the rules does come with some risk (see [IRC](https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2026-03-26#1206466;)), but I don't see any risk with this *specific* change. How to quantify the value of a buggy peer giving a header to a -blocksonly node, I don't know. Not trying to hold this PR hostage, would it be acceptable to split out this change so it gets its own discussion in a smaller, focused PR?

davidgumberg replied:
Agree, sorry i let this sit stale a bit, maybe the right thing to do here since you and @hodlinator seem to agree and I'm willing to concede, is don't process the headers, and if it's wise to actually process the headers that can be discussed/added in a follow up, I feel that it's not important enough of a question to hold this PR up

Crypt-iQ replied:
> and if it's wise to actually process the headers that can be discussed/added in a follow up

This two-step approach has the downside that we're "versioning": any use-case that decides to send compact blocks to -blocksonly on this version, can't, yet they can to the bitcoind version that includes the follow-up. 😅

davidgumberg replied:
well afaict I don't think there's a good reason to do this on purpose, the only reason I would think for processing the headers is in case some buggy implementation sends you a cmpctblock with a valid PoW header that's interesting you want to follow it.

hodlinator reviewed (commented):
Reviewed 82c3a9f6b05a6cd578001da147355fc07d585332

w0xlt reviewed (commented):
reACK 82c3a9f6b05a6cd578001da147355fc07d585332

sedited commented:
@davidgumberg do you want to address the outstanding review comment https://github.com/bitcoin/bitcoin/pull/32606#discussion_r2917119109 ?

davidgumberg commented on test/functional/p2p_compactblocks.py:
```diff
@@ -894,13 +909,17 @@ def announce_cmpct_block(node, peer, txn_count):
             msg = msg_cmpctblock(cmpct_block.to_p2p())
+            peer.send_without_ping(msg_headers([block]))
+            peer.wait_for_getdata([block.hash_int], timeout=30)
+
```
I disagree, this is not added too early. The test will "pass" with these two lines removed, but that is only because the test is not strict enough, consider the following diff applied to e42a906:

```diff
diff --git a/test/functional/p2p_compactblocks.py b/test/functional/p2p_compactblocks.py
index 0d6f8a1e9c..b31d8275a2 100755
--- a/test/functional/p2p_compactblocks.py
+++ b/test/functional/p2p_compactblocks.py
@@ -909,9 +909,10 @@ class CompactBlocksTest(BitcoinTestFramework):
             cmpct_block = HeaderAndShortIDs()
             cmpct_block.initialize_from_block(block)
             msg = msg_cmpctblock(cmpct_block.to_p2p())
-            peer.send_without_ping(msg_headers([block]))
-            peer.wait_for_getdata([block.hash_int], timeout=30)
+            # peer.send_without_ping(msg_headers([block]))
+            # peer.wait_for_getdata([block.hash_int], timeout=30)
 
+            peer.clear_getblocktxn()
             peer.send_and_ping(msg)
             with p2p_lock:
                 assert "getblocktxn" in peer.last_message
```


Until a7895798 the test is not strict enough to actually catch this failure. One could argue that making the test stricter earlier in the commit series is a little bit more commit hygienic, but it's not worth the added churn and IMO increased review difficulty, but if you insist I can add some of these strict checks earlier.

re: adding a boolean so that the solicitation is not always forced as you do in your later commits I'm not sure this adds anything since the tests where this is being used are indifferent to high bandwidth / low-bandwidth status, and the high bandwidth state seems totally incidental at these tests. I think that's a terrible architecture for the test in general as we've discussed above, but this function makes sense to me as "I don't know what HB state the node is in and I don't care, I just want this announcement to be made." which is how the existing tests use it IMO, it just happens to be that some of the nodes they're using already have HB status and don't need the solicit.

hodlinator replied:
> I disagree, this is not added too early.

> One could argue that making the test stricter earlier in the commit series is a little bit more commit hygienic, but it's not worth the added churn and IMO increased review difficulty, but if you insist I can add some of these strict checks earlier.

Making the baseline functional test more strict before the C++ changes in this PR does indeed feel like a better approach. It also makes it clearer why you want to announce the header in `announce_cmpct_block()` early in the commit series.

> re: adding a boolean so that the solicitation is not always forced as you do in your later commits I'm not sure this adds anything since the tests where this is being used are indifferent to high bandwidth / low-bandwidth status, and the high bandwidth state seems totally incidental at these tests. I think that's a terrible architecture for the test in general as we've discussed above, but this function makes sense to me as "I don't know what HB state the node is in and I don't care, I just want this announcement to be made." which is how the existing tests use it IMO, it just happens to be that some of the nodes they're using already have HB status and don't need the solicit.

Regarding earlier discussion on architecture I guess you mean this part from https://github.com/bitcoin/bitcoin/pull/32606#discussion_r2475098091:
> the statefulness of `p2p_compactblocks.py` in general is pretty awful and would be worth fixing in a separate PR.

I agree the shared state between different test cases is awful. However, having precision regarding in which instances we need to coerce a peer into soliciting a block makes for tests that are more likely to catch unintentional/non-obvious behavior changes in the future. A significant part of this PR is about distinguishing solicited and unsolicited compact blocks, so I think tests should also care about that aspect.

---

Experimented with the above here: https://github.com/bitcoin/bitcoin/compare/master...hodlinator:bitcoin:pr/32606_suggestions3
* Starts by a pure refactor commit; merging `announce_cmpct_block()` defs.
* Second commit performs the early tightening up as you suggested, and includes the early headers announcement in `announce_cmpct_block()` which is now required by the tightening up, but wraps it in a `solicited`-param for precision.
* The rest is unchanged except:
  * When `build_compact_block()` is introduced it doesn't add the redundant `assert len(self.utxos)` before the `self.utxos.pop(0)`.
  * We remove `assert len(self.utxos)` in `test_compactblock_reconstruction_parallel_reconstruction()` in the final `qa: Avoid UTXO reuse between test functions` commit instead of in the commit which was merging `announce_cmpct_block()` defs.

davidgumberg replied:
> However, having precision regarding in which instances we need to coerce a peer into soliciting a block makes for tests that are more likely to catch unintentional/non-obvious behavior changes in the future.

My objection is still that whether or not the peers used in these tests are already HB is the arbitrary consequence of the stateful reuse of nodes, it is not an interesting or relevant part of the tests where it's invoked (where it's semantics are as I suggested earlier: have `peer` announce a cmpctblock that is processed) , and I would object to adding such an argument until it was, but I think this is more a matter of style than robustness, and I wouldn't be that upset if you had your way, I'm just not sure it's worth re-churning the PR for.

davidgumberg replied:
Thinking about this more, you are right, I've incorporated something like the changes you suggested.

hodlinator replied:
Thanks for fixing this! Nice to use the block hashes directly instead of `cmpct_block.header.`...

And thanks for scrubbing the unrelated changes from 36d4bdeedfdcbda11178c74c883be6236e67f708, that was embarrassingly messy, good you caught the gist.

davidgumberg commented:
```console
git range-diff 241ad58..82c3a9f 25100fc..5431027
```

I've updated the PR to address @hodlinator and @crypt-iq's feedback from above concerning whether or not headers should be processed in the case of:

1. A peer that has not used `SENDCMPCT` to indicate support for compact blocks.
2. A node that is `-blocksonly`.

Previously, this branch would still process the headers, I have taken their suggestion to drop the block earlier. The question was raised on IRC but no strong input was received. Given what I perceive as the relative insignificance of this choice, I've yielded to the wishes of reviewers here.

I've also rebased in order to fix a silent merge conflict from the removal of `LogIP` in https://github.com/bitcoin/bitcoin/pull/34389/changes/22335474d768f99067856173ff2764b6db753f67.

hodlinator committed:
refactor: Merge announce_cmpct_block() defs into one

hodlinator commented on src/net_processing.cpp:
nanonit: Debatable whether this PR should still be touching these lines as it's just a style change without touching adjacent lines.

davidgumberg replied:
Thanks for catching, fixed.

hodlinator commented on test/functional/p2p_compactblocks.py:
```diff
@@ -946,10 +983,65 @@ def announce_cmpct_block(node, peer, txn_count):
 
+        def build_compact_block():
+            # generate a compact block to send that will force a GETBLOCKTXN
+            assert len(self.utxos)
```
Still not removed.

davidgumberg replied:
Oops, thank you, it's removed now (one hopes!)

hodlinator reviewed (commented):
Reviewed 5431027830e8b0e2ec3cb769a082b62bb7d84cf7

Thanks for bending to our preference of rejecting unsolicited compact blocks before storing the header!

w0xlt reviewed (commented):
reACK 67a65aa860d1c82605ef0c6952e27fb4fea747d5

davidgumberg committed:
test: Tighten getblocktxn checks in parallel cb reconstruction test.

Clear the getblocktxn message so we're not checking existing messages,
and check that the hash in the getblocktxn match the cmpctblock being
announced.

Without this tightening of checks, a later commit that ignores
CMPCTBLOCK messages that are unsolicited will succeed these tests while
silently failing in reality.

Co-authored-by: Hodlinator <172445034+hodlinator@users.noreply.github.com>

hodlinator commented on test/functional/p2p_compactblocks.py:
```diff
@@ -802,21 +852,9 @@ def request_cb_announcements(self, peer):
 
     def test_compactblock_reconstruction_stalling_peer(self, stalling_peer, delivery_peer):
         node = self.nodes[0]
-        assert len(self.utxos)
```
nanonit: This line is currently being removed in 9f5fa626206e6a90797f2285ceaf8761a7c11b92, might be more on-topic to remove it in the final "qa: Avoid UTXO reuse between test functions" commit.

davidgumberg replied:
Thanks for catching this, fixed.

hodlinator commented on test/functional/p2p_compactblocks.py:
```diff
@@ -264,17 +282,37 @@ def check_announcement_of_new_block(node, peer, predicate):
 
     # This test actually causes bitcoind to (reasonably!) disconnect us, so do this last.
+    # Note, these are not consensus-invalid blocks, but improperly constructed cmpctblock messages.
     def test_invalid_cmpctblock_message(self):
```
nit: The 856deb0d431d98253ca9ad968aa808b86061838e "test: (Un)solicited invalid cb -> get disconnected." commit expanding `test_invalid_cmpctblock_message()` is ~fairly~ independent of the C++ changes of the PR. I think it would be better to have it earlier, before a7b3d27cef73d5c9008eaa7e7e6d85dbc355e654 "p2p: Drop unsolicited CMPCTBLOCK from non-HB peer". That way we are more clearly just expanding testing for behavior already existing on master.

Reordered commits which pass p2p_compactblocks.py:
https://github.com/bitcoin/bitcoin/compare/master...hodlinator:bitcoin:pr/32606_suggestions4

davidgumberg replied:
Thanks for the feedback: Your suggestion is better, but I'll avoid reordering the commits for now in hopes of keeping the range-diff small for the update.

hodlinator commented on src/net_processing.cpp:
```diff
@@ -4469,8 +4469,20 @@ void PeerManagerImpl::ProcessMessage(Peer& peer, CNode& pfrom, const std::string
-            LogDebug(BCLog::NET, "Unexpected cmpctblock message received from peer %d\n", pfrom.GetId());
+            LogDebug(BCLog::CMPCTBLOCK, "%s sent us a compact block even though we are still loading blocks!", pfrom.LogPeer());
             return;
+        } else if (m_opts.ignore_incoming_txs) {
+            LogDebug(BCLog::CMPCTBLOCK, "%s sent us a compact block even though we are blocksonly!", pfrom.LogPeer());
+            return;
+        }
+
+        {
+            LOCK(cs_main);
+            const CNodeState *nodestate = State(pfrom.GetId());
+            if (!nodestate->m_provides_cmpctblocks) {
+                LogDebug(BCLog::CMPCTBLOCK, "%s sent us a compact block despite never having sent us a SENDCMPCT!", pfrom.LogPeer());
```
nit: I think the PR description should include a note about these two behavior changes as well (blocks-only + lack of SENDCMPCT).

davidgumberg replied:
Updated the PR description and title, thanks.

hodlinator reviewed (approved):
ACK 59d8f3d7ed3de585e4ff4c6cc9d6c41d6d19921b

* Tightens up the p2p_compactblocks.py test (leaving the statefulness between test cases for another day).
* Changes nodes to drop unsolicited compact blocks from peers we haven't selected as high bandwidth. This decreases the ability to leak information about mempool contents which happens when we respond asking for transactions in the compact block that we do not recognize.
* Adds tests for ignoring such compact blocks.
* Adds tests for us disconnecting peers who send malformed compact blocks (covers existing behavior on master).
* Changes blocks-only nodes drop compact block announcements (since we would have a very empty mempool which in most cases would lack transactions to reconstruct the block).
* Changes nodes to ignore compact blocks from peers which never indicated compact block support.
* Avoids UTXO set reuse in p2p_compactblocks.py (doing this last means we can introduce the constant for how many UTXOs to create with the correct number and avoid having to increase it as tests are added throughout the PR).

Crypt-iQ commented on test/functional/p2p_compactblocks.py:
nit: could assert block hash

Crypt-iQ commented on test/functional/p2p_compactblocks.py:
```diff
@@ -443,6 +447,7 @@ def test_getblocktxn_requests(self, test_node):
+            peer.clear_getblocktxn()
             peer.send_and_ping(msg)
             with p2p_lock:
                 assert "getblocktxn" in peer.last_message
```
nit: assert block hash

Crypt-iQ commented on test/functional/p2p_compactblocks.py:
nit: assert block hash

Crypt-iQ commented on test/functional/p2p_compactblocks.py:
nit: assert block hash, or maybe this could be in some helper

davidgumberg replied:
Thanks, fixed all and added helper.

Crypt-iQ commented on test/functional/p2p_compactblocks.py:
```diff
@@ -906,8 +914,10 @@ def test_compactblock_reconstruction_parallel_reconstruction(self, stalling_peer
+                assert delivery_peer.last_message["getblocktxn"].block_txn_request.blockhash == block.hash_int
             assert_not_equal(node.getbestblockhash(), block.hash_hex)
 
+
```
nit: stray newline

davidgumberg replied:
Thanks, fixed.

Crypt-iQ reviewed (approved):
ACK 59d8f3d7ed3de585e4ff4c6cc9d6c41d6d19921b

left some nits, feel free to address or not

sedited commented:
@davidgumberg do you want to pick up the nits here?

davidgumberg commented:
Pushed to address feedback from @hodlinator and @Crypt-iQ:

- Added `getblocktxn_expected` helper to `p2p_compactblocks.py` with blockhash checks
- Moved the length check removal to the final commit.

Obviously `p2p_compactblocks.py` could be even furthere refactored with a helper that does the clear, send and ping, gbt expected.... but I'd prefer to leave that for a follow up.

w0xlt reviewed (commented):
It should fix the CI errors:

```diff
diff --git a/test/functional/p2p_compactblocks.py b/test/functional/p2p_compactblocks.py
index 3e8655a907..284c465a67 100755
--- a/test/functional/p2p_compactblocks.py
+++ b/test/functional/p2p_compactblocks.py
@@ -158,7 +158,7 @@ class CompactBlocksTest(BitcoinTestFramework):
         assert_equal(gbt.blockhash, blockhash)
         if indices is not None:
             assert_equal(gbt.to_absolute(), indices)
-        assert_not_equal(peer.getbestblockhash(), blockhash)
+        assert_not_equal(int(self.nodes[0].getbestblockhash(), 16), blockhash)
 
     def build_block_on_tip(self, node):
         block = create_block(tmpl=node.getblocktemplate(NORMAL_GBT_REQUEST_PARAMS))
```

davidgumberg commented:
Whoops, pushed a fix to unbreak the test, I always forget that some of the nodes are not real!

w0xlt reviewed (commented):
reACK 55e3a57f22ef9f59ca0bbe90e8aaf4019d8f5783

CI error is unrelated

hodlinator commented on test/functional/p2p_compactblocks.py:
```diff
@@ -150,23 +151,35 @@ def set_test_params(self):
         ]]
         self.utxos = []
 
+    def getblocktxn_expected(self, peer, blockhash, indices=None):
```
nanonit: Realized `self` wasn't used and this together with `assert_highbandwidth_states()` could be hoisted to the outer scope to avoid some tokens.

<details><summary>Diff</summary>

```diff
diff --git a/test/functional/p2p_compactblocks.py b/test/functional/p2p_compactblocks.py
index 73c5066448..dad317e864 100755
--- a/test/functional/p2p_compactblocks.py
+++ b/test/functional/p2p_compactblocks.py
@@ -142,6 +142,27 @@ class TestP2PConn(P2PInterface):
         self.send_without_ping(message)
         self.wait_for_disconnect(timeout=timeout)
 
+
+def getblocktxn_expected(peer, blockhash, indices=None):
+    with p2p_lock:
+        assert "getblocktxn" in peer.last_message
+        gbt = peer.last_message["getblocktxn"].block_txn_request
+
+    assert_equal(gbt.blockhash, blockhash)
+    if indices is not None:
+        assert_equal(gbt.to_absolute(), indices)
+    if isinstance(peer, TestNode):
+        assert_not_equal(peer.getbestblockhash(), blockhash)
+
+
+# assert the RPC getpeerinfo boolean fields `bip152_hb_{to, from}`
+# match the given parameters for the peer at idx of a given node
+def assert_highbandwidth_states(node, idx, hb_to, hb_from):
+    peerinfo = node.getpeerinfo()[idx]
+    assert_equal(peerinfo['bip152_hb_to'], hb_to)
+    assert_equal(peerinfo['bip152_hb_from'], hb_from)
+
+
 class CompactBlocksTest(BitcoinTestFramework):
     def set_test_params(self):
         self.setup_clean_chain = True
@@ -151,17 +172,6 @@ class CompactBlocksTest(BitcoinTestFramework):
         ]]
         self.utxos = []
 
-    def getblocktxn_expected(self, peer, blockhash, indices=None):
-        with p2p_lock:
-            assert "getblocktxn" in peer.last_message
-            gbt = peer.last_message["getblocktxn"].block_txn_request
-
-        assert_equal(gbt.blockhash, blockhash)
-        if indices is not None:
-            assert_equal(gbt.to_absolute(), indices)
-        if isinstance(peer, TestNode):
-            assert_not_equal(peer.getbestblockhash(), blockhash)
-
     def build_block_on_tip(self, node):
         block = create_block(tmpl=node.getblocktemplate(NORMAL_GBT_REQUEST_PARAMS))
         block.solve()
@@ -203,7 +213,7 @@ class CompactBlocksTest(BitcoinTestFramework):
 
         peer.clear_getblocktxn()
         peer.send_and_ping(msg)
-        self.getblocktxn_expected(peer, block.hash_int)
+        getblocktxn_expected(peer, block.hash_int)
         return block, cmpct_block
 
     # Test "sendcmpct" (between peers preferring the same version):
@@ -298,12 +308,12 @@ class CompactBlocksTest(BitcoinTestFramework):
         hb_peer = self.nodes[0].add_p2p_connection(TestP2PConn())
         self.request_cb_announcements(hb_peer)
         self.make_peer_hb_to_candidate(self.nodes[0], hb_peer)
-        self.assert_highbandwidth_states(self.nodes[0], idx=-1, hb_to=True, hb_from=True)
+        assert_highbandwidth_states(self.nodes[0], idx=-1, hb_to=True, hb_from=True)
 
         # Make a low-bandwidth peer
         lb_peer = self.nodes[0].add_p2p_connection(TestP2PConn())
         self.request_cb_announcements(lb_peer)
-        self.assert_highbandwidth_states(self.nodes[0], idx=-1, hb_to=False, hb_from=True)
+        assert_highbandwidth_states(self.nodes[0], idx=-1, hb_to=False, hb_from=True)
 
         # Construct an invalid cmpctblock message
         self.generate(self.nodes[0], COINBASE_MATURITY + 1)
@@ -446,7 +456,7 @@ class CompactBlocksTest(BitcoinTestFramework):
             test_node.send_and_ping(msg_cmpctblock(comp_block.to_p2p()))
             assert_equal(int(node.getbestblockhash(), 16), block.hashPrevBlock)
             # Expect a getblocktxn message that requests the coinbase.
-            self.getblocktxn_expected(test_node, block.hash_int, indices=[0])
+            getblocktxn_expected(test_node, block.hash_int, indices=[0])
 
             # Send the coinbase, and verify that the tip advances.
             msg = msg_blocktxn()
@@ -480,7 +490,7 @@ class CompactBlocksTest(BitcoinTestFramework):
             msg = msg_cmpctblock(compact_block.to_p2p())
             peer.clear_getblocktxn()
             peer.send_and_ping(msg)
-            self.getblocktxn_expected(peer, compact_block.header.hash_int, expected_result)
+            getblocktxn_expected(peer, compact_block.header.hash_int, expected_result)
 
         def test_tip_after_message(node, peer, msg, tip):
             peer.send_and_ping(msg)
@@ -573,7 +583,7 @@ class CompactBlocksTest(BitcoinTestFramework):
         test_node.clear_getblocktxn()
         test_node.send_and_ping(msg_cmpctblock(comp_block.to_p2p()))
         expected_indices = [6, 7, 8, 9, 10]
-        self.getblocktxn_expected(test_node, block.hash_int, expected_indices)
+        getblocktxn_expected(test_node, block.hash_int, expected_indices)
 
         # Now give an incorrect response.
         # Note that it's possible for bitcoind to be smart enough to know we're
@@ -614,7 +624,7 @@ class CompactBlocksTest(BitcoinTestFramework):
         comp_block.initialize_from_block(block, prefill_list=[0], use_witness=True)
         test_node.clear_getblocktxn()
         test_node.send_and_ping(msg_cmpctblock(comp_block.to_p2p()))
-        self.getblocktxn_expected(test_node, block.hash_int, indices=[1,2])
+        getblocktxn_expected(test_node, block.hash_int, indices=[1,2])
 
         # Send a blocktxn that does not succeed in reconstruction, triggering
         # getdata fallback.
@@ -885,14 +895,6 @@ class CompactBlocksTest(BitcoinTestFramework):
         stalling_peer.send_and_ping(msg)
         assert_equal(node.getbestblockhash(), block.hash_hex)
 
-    # assert the RPC getpeerinfo boolean fields `bip152_hb_{to, from}`
-    # match the given parameters for the peer at idx of a given node
-    @staticmethod
-    def assert_highbandwidth_states(node, idx, hb_to, hb_from):
-        peerinfo = node.getpeerinfo()[idx]
-        assert_equal(peerinfo['bip152_hb_to'], hb_to)
-        assert_equal(peerinfo['bip152_hb_from'], hb_from)
-
     def test_highbandwidth_mode_states_via_getpeerinfo(self):
         # create new p2p connection for a fresh state w/o any prior sendcmpct messages sent
         hb_test_node = self.nodes[0].add_p2p_connection(TestP2PConn())
@@ -901,21 +903,21 @@ class CompactBlocksTest(BitcoinTestFramework):
         hb_test_node_idx = -1
 
         # initially, neither node has selected the other peer as high-bandwidth yet
-        self.assert_highbandwidth_states(self.nodes[0], hb_test_node_idx, hb_to=False, hb_from=False)
+        assert_highbandwidth_states(self.nodes[0], hb_test_node_idx, hb_to=False, hb_from=False)
 
         # peer requests high-bandwidth mode by sending sendcmpct(1)
         hb_test_node.send_and_ping(msg_sendcmpct(announce=True, version=2))
-        self.assert_highbandwidth_states(self.nodes[0], hb_test_node_idx, hb_to=False, hb_from=True)
+        assert_highbandwidth_states(self.nodes[0], hb_test_node_idx, hb_to=False, hb_from=True)
 
         # peer generates a block and sends it to node, which should
         # select the peer as high-bandwidth (up to 3 peers according to BIP 152)
         block = self.build_block_on_tip(self.nodes[0])
         hb_test_node.send_and_ping(msg_block(block))
-        self.assert_highbandwidth_states(self.nodes[0], hb_test_node_idx, hb_to=True, hb_from=True)
+        assert_highbandwidth_states(self.nodes[0], hb_test_node_idx, hb_to=True, hb_from=True)
 
         # peer requests low-bandwidth mode by sending sendcmpct(0)
         hb_test_node.send_and_ping(msg_sendcmpct(announce=False, version=2))
-        self.assert_highbandwidth_states(self.nodes[0], hb_test_node_idx, hb_to=True, hb_from=False)
+        assert_highbandwidth_states(self.nodes[0], hb_test_node_idx, hb_to=True, hb_from=False)
 
     def test_compactblock_reconstruction_parallel_reconstruction(self, stalling_peer, delivery_peer, inbound_peer, outbound_peer):
         """ All p2p connections are inbound except outbound_peer. We test that ultimate parallel slot
@@ -947,7 +949,7 @@ class CompactBlocksTest(BitcoinTestFramework):
 
             delivery_peer.send_and_ping(msg_cmpctblock(cmpct_block.to_p2p()))
             # The second peer to announce should still get a getblocktxn
-            self.getblocktxn_expected(delivery_peer, block.hash_int)
+            getblocktxn_expected(delivery_peer, block.hash_int)
 
             inbound_peer.send_and_ping(msg_cmpctblock(cmpct_block.to_p2p()))
             with p2p_lock:
@@ -957,7 +959,7 @@ class CompactBlocksTest(BitcoinTestFramework):
 
             outbound_peer.send_and_ping(msg_cmpctblock(cmpct_block.to_p2p()))
             # The third peer to announce should get a getblocktxn if outbound
-            self.getblocktxn_expected(outbound_peer, block.hash_int)
+            getblocktxn_expected(outbound_peer, block.hash_int)
 
             # Second peer completes the compact block first
             msg = msg_blocktxn()
@@ -998,7 +1000,7 @@ class CompactBlocksTest(BitcoinTestFramework):
 
         # create new p2p connection for a fresh state w/o any prior sendcmpct messages sent
         unsolicited_peer = self.nodes[0].add_p2p_connection(TestP2PConn())
-        self.assert_highbandwidth_states(node, idx=-1, hb_to=False, hb_from=False)
+        assert_highbandwidth_states(node, idx=-1, hb_to=False, hb_from=False)
 
         self.log.info("Test that a node ignores unsolicited CMPCTBLOCK messages from peers that have not sent SENDCMPCT.")
         assert ignores_compact_block(unsolicited_peer, solicited=False)
@@ -1009,15 +1011,15 @@ class CompactBlocksTest(BitcoinTestFramework):
         # Unsolicited peer announces CMPCTBLOCK support with SENDCMPCT message,
         # but still non-HB.
         unsolicited_peer.send_and_ping(msg_sendcmpct())
-        self.assert_highbandwidth_states(node, idx=-1, hb_to=False, hb_from=False)
+        assert_highbandwidth_states(node, idx=-1, hb_to=False, hb_from=False)
 
         self.log.info("Test that a node ignores unsolicited CMPCTBLOCK messages from non-HB peers.")
         assert ignores_compact_block(unsolicited_peer, solicited=False)
-        self.assert_highbandwidth_states(node, idx=-1, hb_to=False, hb_from=False)
+        assert_highbandwidth_states(node, idx=-1, hb_to=False, hb_from=False)
 
         self.log.info("Test that a node does not ignore solicited CMPCTBLOCK messages from non-HB peers.")
         assert not ignores_compact_block(unsolicited_peer, solicited=True)
-        self.assert_highbandwidth_states(node, idx=-1, hb_to=False, hb_from=False)
+        assert_highbandwidth_states(node, idx=-1, hb_to=False, hb_from=False)
 
         # The node will ask for transactions from an unsolicited compact block
         # it receives from a high bandwidth peer, we need to use one set up earlier,
@@ -1025,7 +1027,7 @@ class CompactBlocksTest(BitcoinTestFramework):
         self.log.info("Test that a node does not ignore unsolicited CMPCTBLOCK messages from HB peers.")
 
         hb_peer_idx = -2
-        self.assert_highbandwidth_states(node, idx=hb_peer_idx, hb_to=True, hb_from=False)
+        assert_highbandwidth_states(node, idx=hb_peer_idx, hb_to=True, hb_from=False)
         hb_peer = self.nodes[0].p2ps[hb_peer_idx]
         assert not ignores_compact_block(hb_peer, solicited=False)
 
```

</details>

davidgumberg replied:
Good catch, will address if retouching.

hodlinator reviewed (approved):
re-ACK 55e3a57f22ef9f59ca0bbe90e8aaf4019d8f5783

New `getblocktxn_expected()` helps clean things up. 👍

sedited commented:
@polespinasa @w0xlt mind giving this another look?

polespinasa commented on src/net_processing.cpp:
```diff
@@ -2610,7 +2610,7 @@ void PeerManagerImpl::SendBlockTransactions(CNode& pfrom, Peer& peer, const CBlo
         uint32_t tx_requested_size{0};
         for (const auto& tx : resp.txn) tx_requested_size += tx->ComputeTotalSize();
-        LogDebug(BCLog::CMPCTBLOCK, "Peer %d sent us a GETBLOCKTXN for block %s, sending a BLOCKTXN with %u txns. (%u bytes)\n", pfrom.GetId(), block.GetHash().ToString(), resp.txn.size(), tx_requested_size);
+        LogDebug(BCLog::CMPCTBLOCK, "%s sent us a GETBLOCKTXN for block %s, sending a BLOCKTXN with %u txns. (%u bytes)", pfrom.LogPeer(), block.GetHash().ToString(), resp.txn.size(), tx_requested_size);
```
in 831359171582fc8220f9082049fddc0cc185f93d p2p: Drop unsolicited CMPCTBLOCK from non-HB peer

Ignore this nit, for sure not worth at the current state of the PR, but if re-touching probably this change can be done in a separate commit.

davidgumberg replied:
Thanks for catching, I'll address if retouching.

polespinasa commented on test/functional/p2p_compactblocks.py:
```diff
@@ -167,18 +167,19 @@ def build_block_on_tip(self, node):
         return block
 
-    # Create 10 more anyone-can-spend utxo's for testing.
+    # Create 12 more anyone-can-spend utxo's for testing.
```
in 55e3a57f22ef9f59ca0bbe90e8aaf4019d8f5783 qa: Avoid UTXO reuse between test functions

Why change it from 10 to 12?

davidgumberg replied:
Good question, prior to this commit, some utxo's were reused between tests e.g. `test_multiple_blocktxn_response` uses a utxo but never pops it. This commit fixes all the spots where utxo's are reused by `pop()`ing them, which means more UTXO's are needed in the pool for all the tests. If you try running the test with `COUNT = 10` you will see the issue.

polespinasa replied:
oh I missed, that makes sense

polespinasa commented on test/functional/p2p_compactblocks.py:
```diff
@@ -150,6 +151,17 @@ def set_test_params(self):
+        assert_equal(gbt.blockhash, blockhash)
+        if indices is not None:
+            assert_equal(gbt.to_absolute(), indices)
+        if isinstance(peer, TestNode):
```
in 25457a3272ff3b4ea6d30503f725675a8d16b987 test: Tighten getblocktxn checks in parallel cb reconstruction test.

I think this is always falsy, peer is always a `TestP2PConn` not a `TestNode`.
The assert below is dead code and the `TestNode` include can be dropped.

davidgumberg replied:
Seems reasonable, but given that the code is correct even if it's not executed I think it's fine to leave as is. It also acts as a bit of documentation for why this check doesn't apply to many of the existing tests when one might assume that it does.

polespinasa commented on src/net_processing.cpp:
```diff
@@ -4469,7 +4469,10 @@ void PeerManagerImpl::ProcessMessage(Peer& peer, CNode& pfrom, const std::string
         // Ignore cmpctblock received while importing
         if (m_chainman.m_blockman.LoadingBlocks()) {
-            LogDebug(BCLog::NET, "Unexpected cmpctblock message received from peer %d\n", pfrom.GetId());
+            LogDebug(BCLog::CMPCTBLOCK, "%s sent us a compact block even though we are still loading blocks!", pfrom.LogPeer());
```
in bf9884f4e55df502b67b2636969cacce62edaee9 p2p: make blocksonly nodes ignore CMPCTBLOCK messages

nit: `!` looks weird imho

davidgumberg replied:
Exclamation points look awesome, and do a great job of communicating how shocking this information is.

polespinasa reviewed (commented):
lgtm ACK 55e3a57f22ef9f59ca0bbe90e8aaf4019d8f5783

Left some really small nits

achow101 commented:
ACK 55e3a57f22ef9f59ca0bbe90e8aaf4019d8f5783
